sfge/src/test/java/com/salesforce/rules/UnnecessarilyExtensibleClassRuleTest.java

```diff
@@ -0,0 +1,188 @@
+package com.salesforce.rules;
+
+import static org.hamcrest.Matchers.hasSize;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import com.salesforce.TestUtil;
+import java.util.*;
+import java.util.stream.Stream;
+import org.apache.tinkerpop.gremlin.process.traversal.dsl.graph.GraphTraversalSource;
+import org.hamcrest.MatcherAssert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class UnnecessarilyExtensibleClassRuleTest {
+    private GraphTraversalSource g;
+
+    @BeforeEach
+    public void setup() {
+        this.g = TestUtil.getGraph();
+    }
+
+    @MethodSource("paramProvider_testOuterClasses")
+    @ParameterizedTest(name = "{displayName}: Scope {0}")
+    public void testOuterClasses(String scope, Set<String> expectedDefiningTypes) {
+        // This class is never extended.
+        String unextendedSource = scope + " class UnextendedClass {}";
+        // This class is extended.
+        String extendedSource = scope + " class ExtendedClass {}";
+        // This class does the extension.
+        String extenderSource = "public class ExtenderClass extends ExtendedClass {}";
+
+        TestUtil.buildGraph(g, unextendedSource, extendedSource, extenderSource);
+
+        StaticRule rule = UnnecessarilyExtensibleClassRule.getInstance();
+        List<Violation> violations = rule.run(g);
+
+        MatcherAssert.assertThat(
+                "Wrong number of violations found",
+                violations,
+                hasSize(expectedDefiningTypes.size()));
+        for (int i = 0; i < expectedDefiningTypes.size(); i++) {
+            String actual = violations.get(i).getSourceDefiningType();
+            assertTrue(
+                    expectedDefiningTypes.contains(actual),
+                    String.format("%s should not be unextended", actual));
```

Comment:
Once we have the string constants moved to UserFacingMessages.java, can we refer to the actual constant so that the test doesn't have to be modified for text changes?

Reply:
Same for other tests with `String.format` too.

Reply:
The `String.format` call here is to properly format a message for test failure, not to check what the error's message is.